File: backend/app/schemas/reports.py
"""
Reporting Schemas
"""
from datetime import datetime
from typing import List, Dict, Optional
from pydantic import BaseModel, Field

from ..models.prediction import DiseaseType, RiskLevel


class ClinicalPatientSummary(BaseModel):
    id: int
    patient_identifier: str
    full_name: str
    age: float = Field(..., ge=0)
    gender: str


class ClinicalPredictionSummary(BaseModel):
    id: int
    created_at: datetime
    disease_type: DiseaseType
    alzheimer_risk_score: Optional[float] = Field(None, ge=0, le=1)
    alzheimer_risk_level: Optional[RiskLevel] = None
    parkinson_risk_score: Optional[float] = Field(None, ge=0, le=1)
    parkinson_risk_level: Optional[RiskLevel] = None
    recommendations: Optional[str] = None


class ClinicalReport(BaseModel):
    patient: ClinicalPatientSummary
    predictions: List[ClinicalPredictionSummary]
    last_medical_record_at: Optional[datetime] = None
    pending_follow_up: bool = False


class ResearchAggregation(BaseModel):
    disease_type: DiseaseType
    risk_level: Optional[RiskLevel] = None
    count: int


class ResearchReport(BaseModel):
    total_predictions: int
    unique_patients: int
    aggregation: List[ResearchAggregation]
    timeframe_start: Optional[datetime] = None
    timeframe_end: Optional[datetime] = None


class ManagementKpi(BaseModel):
    total_predictions: int
    reviewed_predictions: int
    active_patients: int
    avg_response_time_ms: Optional[float] = None


class ManagementAlert(BaseModel):
    title: str
    severity: str
    description: str
    created_at: datetime


class ManagementReport(BaseModel):
    kpi: ManagementKpi
    model_version_distribution: Dict[str, int]
    alerts: List[ManagementAlert]


class ReportExportRequest(BaseModel):
    report_type: str
    format: str = Field(..., pattern="^(pdf|excel|csv)$")
    filters: Dict[str, Optional[str]] = Field(default_factory=dict)


class ReportExportResponse(BaseModel):
    message: str
    report_type: str
    format: str
    filters: Dict[str, Optional[str]]
    generated_at: datetime

File: admin-dashboard/src/services/reports.ts
import axios from 'axios'

const API_BASE = '/api/v1/reports'

export type ClinicalReportFilters = {
  patientId: number
  start?: string
  end?: string
}

export type ResearchReportFilters = {
  start?: string
  end?: string
  diseaseType?: string
  riskLevel?: string
}

export type ManagementReportFilters = {
  modelVersion?: string
  start?: string
  end?: string
}

export type ClinicalReport = {
  patient: {
    id: number
    patient_identifier: string
    full_name: string
    age: number
    gender: string
  }
  predictions: Array<{
    id: number
    created_at: string
    disease_type: string
    alzheimer_risk_score?: number
    alzheimer_risk_level?: string
    parkinson_risk_score?: number
    parkinson_risk_level?: string
    recommendations?: string
  }>
  last_medical_record_at?: string | null
  pending_follow_up: boolean
}

export type ResearchReport = {
  total_predictions: number
  unique_patients: number
  aggregation: Array<{
    disease_type: string
    risk_level?: string | null
    count: number
  }>
  timeframe_start?: string | null
  timeframe_end?: string | null
}

export type ManagementReport = {
  kpi: {
    total_predictions: number
    reviewed_predictions: number
    active_patients: number
    avg_response_time_ms?: number | null
  }
  model_version_distribution: Record<string, number>
  alerts: Array<{
    title: string
    severity: string
    description: string
    created_at: string
  }>
}

export type ReportExportResponse = {
  message: string
  report_type: string
  format: string
  generated_at: string
}

const buildParams = (filters: Record<string, unknown>) => {
  const params = new URLSearchParams()
  Object.entries(filters).forEach(([key, value]) => {
    if (value !== undefined && value !== null && value !== '') {
      params.append(key, String(value))
    }
  })
  return params
}

export const reportsService = {
  async fetchClinical(filters: ClinicalReportFilters) {
    const params = buildParams({
      patient_id: filters.patientId,
      start: filters.start,
      end: filters.end,
    })
    const response = await axios.get(`${API_BASE}/clinical?${params.toString()}`)
    return response.data as ClinicalReport
  },

  async fetchResearch(filters: ResearchReportFilters) {
    const params = buildParams({
      start: filters.start,
      end: filters.end,
      disease_type: filters.diseaseType,
      risk_level: filters.riskLevel,
    })
    const response = await axios.get(`${API_BASE}/research?${params.toString()}`)
    return response.data as ResearchReport
  },

  async fetchManagement(filters: ManagementReportFilters) {
    const params = buildParams({
      model_version: filters.modelVersion,
      start: filters.start,
      end: filters.end,
    })
    const response = await axios.get(`${API_BASE}/management?${params.toString()}`)
    return response.data as ManagementReport
  },

  async exportReport(payload: { reportType: string; format: 'pdf' | 'excel' | 'csv'; filters: Record<string, unknown> }) {
    const response = await axios.post(`${API_BASE}/export`, {
      report_type: payload.reportType,
      format: payload.format,
      filters: payload.filters,
    })
    return response.data as ReportExportResponse
  },
}


